pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"context"
+	"sync/atomic"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+	schedutil "k8s.io/kubernetes/pkg/scheduler/util"
+
+	"k8s.io/klog"
+)
+
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type topologySpreadConstraintsMap struct {
+	// podCounts is keyed with node name, and valued with the number of matching pods.
+	podCounts map[string]*int64
+	// total number of matching pods on each qualified <topologyKey:value> pair
+	total int64
+	// topologyPairToNodeNames store the mapping from potential <topologyKey:value>
+	// pair to node names
+	topologyPairToNodeNames map[topologyPair][]string
+}
+
+func newTopologySpreadConstraintsMap(len int) *topologySpreadConstraintsMap {
+	return &topologySpreadConstraintsMap{
+		podCounts:               make(map[string]*int64, len),
+		topologyPairToNodeNames: make(map[topologyPair][]string),
+	}
+}
+
+func (t *topologySpreadConstraintsMap) initialize(pod *v1.Pod, nodes []*v1.Node) {
+	constraints := getSoftTopologySpreadConstraints(pod)
+	for _, node := range nodes {
+		match := true
+		var pairs []topologyPair
+		for _, constraint := range constraints {
+			tpKey := constraint.TopologyKey
+			if _, ok := node.Labels[tpKey]; !ok {
+				// Current node isn't qualified for the soft constraints,
+				// so break here and the node will hold default value (nil).
+				match = false
+				break
+			}
+			pairs = append(pairs, topologyPair{key: tpKey, value: node.Labels[tpKey]})
+		}
+		if match {
+			for _, pair := range pairs {
+				t.topologyPairToNodeNames[pair] = append(t.topologyPairToNodeNames[pair], node.Name)
```

Comment:
You can avoid these lists doing the following:
For each topology pair, calculate how many pods match `O(P*C)`. Then iterate through the nodes and sum up the counters for each of its topologies `O(N*C)`. Total `O(P*C)`.

P=# pods, C=# constraints, N=# nodes.

At first I thought your algorithm was `O(N^2*C)`, because you iterate through nodes within your parallel for. But it actually amortizes to `O(P*C)` plus the same `O(N*C)`, because a node only belongs to one set of topologies. In any case, you can get rid of the per-node map with a per-topologypair map. And sum up later for each node.

This can all be done in a single parallel for on the nodes, if you use appropriate synchronization primitives (like wait group).

Reply:
Per our offline talk, we should note that `N` can be `N1` (all nodes) and `N2` (filtered nodes which passed the Predicates). So generally speaking, no algorithm can do better than `N1` * `P` (P = pods in **one** node) * `C`. I put an additional `N2` loop in `initilize` is to filter out non-qualified nodes, as well as do some prepping data structures.

Please check out the latest version. It also includes some benchmark tests. It's doing a much better job.

```
## Old Commit

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     300	   3880228 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   4121098 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   4084106 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	     500	   3704594 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	     300	   5029568 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	     500	   3408629 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     200	   7106261 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     200	   7428552 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     200	   6032280 ns/op

## New Commit

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   3133159 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   2814902 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-zone-8         	     500	   3206365 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	    1000	   1802340 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	    1000	   1869673 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/single-constraint-node-8         	    1000	   1927666 ns/op

BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     300	   4197458 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     300	   4181693 ns/op
BenchmarkTestCalculateEvenPodsSpreadPriority/1000nodes/two-constraints-zone-node-8      	     300	   4242142 ns/op
```

Reply:
Btw, this looks awesome.